File: src/main/java/com/janrain/backplane/server2/oauth2/AuthorizationRequest.java
/*
 * Copyright 2012 Janrain, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.janrain.backplane.server2.oauth2;

import com.janrain.backplane.common.BackplaneServerException;
import com.janrain.backplane.common.DateTimeUtils;
import com.janrain.backplane.common.ExternalizableCore;
import com.janrain.backplane.server2.Client;
import com.janrain.backplane.server2.dao.ClientDAO;
import com.janrain.backplane.servlet.InvalidRequestException;
import com.janrain.commons.message.MessageException;
import com.janrain.commons.message.MessageField;
import org.apache.commons.lang.StringUtils;

import java.util.*;

/**
 * @author Johnny Bufu
 */
public class AuthorizationRequest extends ExternalizableCore {

    // - PUBLIC

    /**
     * Empty default constructor for AWS to use
     */
    public AuthorizationRequest() { }

    public AuthorizationRequest(String cookie, Map parameterMap) throws MessageException {
        Map<String,String> data = new LinkedHashMap<String, String>();
        for(Field f: EnumSet.allOf(Field.class)) {
            Object value = parameterMap.get(f.getFieldName().toLowerCase());
            if ( value != null && ((String[]) value).length > 0 && StringUtils.isNotEmpty(((String[])value)[0])) {
                data.put(f.getFieldName(), ((String[])value)[0]);
            }
        }
        data.put(Field.COOKIE.getFieldName(), cookie);
        data.put(Field.EXPIRES.getFieldName(), DateTimeUtils.ISO8601.get().format(new Date(System.currentTimeMillis() + AUTH_REQUEST_TIMEOUT_SECONDS * 1000)));
        super.init(cookie, data);
    }

    @Override
    public String getIdValue() {
        return get(Field.COOKIE);
    }

    @Override
    public Set<? extends MessageField> getFields() {
        return EnumSet.allOf(Field.class);
    }
    
    public String getRedirectUri(ClientDAO clientDAO) throws AuthorizationException {
        String client_id = get(Field.CLIENT_ID);

        Client client = null;
        try {
            client = clientDAO.get(client_id);
        } catch (BackplaneServerException e) {
            // ignore
        }
        if(client == null) {
            throw new AuthorizationException(OAuth2.OAUTH2_AUTHZ_DIRECT_ERROR, "invalid client_id: " + client_id , this);
        }
        String requestRedirectUri = get(Field.REDIRECT_URI);

        if (StringUtils.isNotEmpty(requestRedirectUri)) {
            try {
                OAuth2.validateRedirectUri(requestRedirectUri, client.get(Client.ClientField.REDIRECT_URI));
            } catch (ValidationException ve) {
                throw new AuthorizationException(OAuth2.OAUTH2_AUTHZ_DIRECT_ERROR, "invalid redirect_uri: " + ve.getMessage() , this);
            }
            return requestRedirectUri;
        } else {
            return client.get(Client.ClientField.REDIRECT_URI);
        }
    }

    public static enum Field implements MessageField {

        // - PUBLIC

        COOKIE,
        EXPIRES,
        CLIENT_ID,
        RESPONSE_TYPE {
            @Override
            public void validate(String value) throws MessageException {
                super.validate(value);
                if ( ! OAuth2.OAUTH2_TOKEN_RESPONSE_TYPE_CODE.equals(value)) {
                    throw new IllegalArgumentException("Unsupported OAuth2 response_type: " + value);
                }
            }
        },
        REDIRECT_URI(false) {
            @Override
            public void validate(String value) throws MessageException {
                super.validate(value);
                try {
                    OAuth2.validateRedirectUri(value);
                } catch (ValidationException e) {
                    throw new InvalidRequestException(e.getMessage());
                }
            }
        },
        SCOPE(false),
        STATE(false);

        @Override
        public String getFieldName() {
            return name();
        }

        @Override
        public boolean isRequired() {
            return required;
        }

        @Override
        public void validate(String value) throws MessageException {
            if (isRequired()) validateNotBlank(getFieldName().toLowerCase(), value);
        }

        // - PRIVATE

        private final boolean required;

        private Field() {
            this(true);
        }

        private Field(boolean required) {
            this.required = required;
        }
    }

    // - PRIVATE

    private static final long serialVersionUID = 5105927572907870325L;

    private static final long AUTH_REQUEST_TIMEOUT_SECONDS = 1200l;
}

File: src/main/scala/com/janrain/backplane/server2/oauth2/model/AuthorizationRequest.scala
package com.janrain.backplane.server2.oauth2.model

import com.janrain.oauth2.{AuthorizationException, ValidationException, OAuth2}
import com.janrain.servlet.InvalidRequestException
import org.apache.commons.lang.exception.ExceptionUtils
import com.janrain.backplane.common.DateTimeUtils
import java.util.Date
import com.janrain.backplane.common.model.{MessageField, MessageFieldEnum, Message}
import com.janrain.backplane.server2.dao.BP2DAOs

/**
 * @author Johnny Bufu
 */
class AuthorizationRequest(data: Map[String,String]) extends Message(data, AuthorizationRequestFields.values) {

  def this(cookie: String, data: Map[String,String]) = this(data ++
    Map( AuthorizationRequestFields.COOKIE.name -> cookie,
         AuthorizationRequestFields.EXPIRES.name
           -> DateTimeUtils.ISO8601.get().format(new Date(System.currentTimeMillis() + AuthorizationRequest.AUTH_REQUEST_TIMEOUT_SECONDS * 1000))
    ))

  def idField = AuthorizationRequestFields.COOKIE

  def getRedirectUri: String = {
    val client_id = get(AuthorizationRequestFields.CLIENT_ID)
    client_id
      .map(BP2DAOs.clientDao.get(_))
      .map(validRedirectUri(_, get(AuthorizationRequestFields.REDIRECT_URI)))
      .getOrElse(
       throw new AuthorizationException(OAuth2.OAUTH2_AUTHZ_DIRECT_ERROR, "invalid client_id: " + client_id, this)
      )
  }

  private def validRedirectUri(client: Option[Client], requestRedirectUri: Option[String]): String = {
    val clientRedirectUri = client.flatMap(_.get(ClientFields.REDIRECT_URI))
    (for {
      configuredUri <- clientRedirectUri
      requestedUri <- requestRedirectUri
    } yield {
      try {
        OAuth2.validateRedirectUri(requestedUri, configuredUri)
        requestedUri
      } catch {
        case e: ValidationException =>
          throw new AuthorizationException(OAuth2.OAUTH2_AUTHZ_DIRECT_ERROR, "invalid redirect_uri: " + e.getMessage, this)
      }
    })
    .getOrElse(clientRedirectUri.getOrElse(
      throw new AuthorizationException(OAuth2.OAUTH2_AUTHZ_DIRECT_ERROR, "no redirect_uri configured for client_id: " + id, this)))
  }
}

object AuthorizationRequest {
  private[model] final val AUTH_REQUEST_TIMEOUT_SECONDS: Long = 1200L
}

object AuthorizationRequestFields extends MessageFieldEnum {

  type AuthorizationRequestField = EnumVal

  sealed trait EnumVal extends Value with MessageField {
    def required = true
  }

  val COOKIE = new AuthorizationRequestField { def name = "cookie" }

  val EXPIRES = new AuthorizationRequestField { def name = "expires" }

  val CLIENT_ID = new AuthorizationRequestField { def name = "client_id" }

  val RESPONSE_TYPE  = new AuthorizationRequestField { def name = "response_type"

    override def validate(fieldValue: Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      if (fieldValue.exists(_ != OAuth2.OAUTH2_TOKEN_RESPONSE_TYPE_CODE)) {
        throw new IllegalArgumentException("Unsupported OAuth2 response_type: " + fieldValue)
      }
    }
  }

  val REDIRECT_URI = new AuthorizationRequestField { def name = "redirect_uri"; override def required = false

    override def validate(fieldValue: Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      fieldValue.foreach(
        try {
          OAuth2.validateRedirectUri(_)
        } catch {
          case e: ValidationException => throw new InvalidRequestException(ExceptionUtils.getRootCauseMessage(e))
        })
    }
  }

  val SCOPE = new AuthorizationRequestField { def name = "scope"; override def required = false }

  val STATE = new AuthorizationRequestField { def name = "state"; override def required = false }
}